torch/nn/modules/transformer.py

```diff
@@ -150,11 +165,15 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
         return output
 
     @staticmethod
-    def generate_square_subsequent_mask(sz: int, device='cpu') -> Tensor:
-        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
+    def generate_square_subsequent_mask(
+            sz: int,
+            device: torch.device = torch.device('cpu'),
```

Comment:
Is there a reason why historically we always default to creating the mask on cpu here?

For context `device` and `dtype` are usually factory kwargs that default to None (e.g. [here](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L57-L58))
e.g.
```
import torch
factory_kwargs = {'device': None, 'dtype': None}
sz = 2
t = torch.full((sz, sz), float('-inf'), **factory_kwargs)
print(t.dtype, t.device)
```

 It would be nice if we could do the same here but if we make this change here though it will break BC for the edge case where the user did `torch.set_default_device('cuda')` outside the function. So I am open to leaving this like how you have it.

Reply:
I tried permutations of thisI'm very receptive to this.  I tried with None and ran into the type checker issues -- I don't know if there's a straight-forward way.  that's why I have torch.get_default_dtype() on the dtype. Maybe we can do the same for device.

Reply:
I found a way to do this by using `torch.device(torch.C._get_default_device())` --- wayyyy to hide this functionality!

:)